### PR TITLE
Eliminate time-consuming LocalMergeSortSourceOperator and parallelize AnalyticNode with partition_exprs

### LINKED ISSUE #1940: Eliminate time-consuming LocalMergeSortSourceOperator and parallelize AnalyticNode with partition_exprs
## Enhancement
In pipeline execution engine,  PartitionSortSinkOperator and LocalMergeSortSourceOperator are prepended to AnalyticNode with partition_exprs. if PartitionSortSinkOperator's dop >1, then LocalMergeSortSourceOperator merges multiple sorted input streams output produced by PartitionSortSinkOperator instances and the merging operation is quite time-consuming, so a LocalShuffle operator with partition_exprs as the same as counterpart of  AnalyticNode is prepended to PartitionSortSinkOperators, then each PartitionSortSinkOperator has an independent LocalMergeSortSourceOperator, so the merging operation is eliminated, AnalyticNode with partition_exprs now can parallelized, each AnalyticSinkOperator processes an sorted partition.

### PR DESCRIPTION
**Pull request by satanson**
For https://github.com/StarRocks/starrocks/issues/1940

## Enhancement
In pipeline execution engine,  PartitionSortSinkOperator and LocalMergeSortSourceOperator are prepended to AnalyticNode with partition_exprs. if PartitionSortSinkOperator's dop >1, then LocalMergeSortSourceOperator merges multiple sorted input streams output produced by PartitionSortSinkOperator instances and the merging operation is quite time-consuming, so a LocalShuffle operator with partition_exprs as the same as counterpart of  AnalyticNode is prepended to PartitionSortSinkOperators, then each PartitionSortSinkOperator has an independent LocalMergeSortSourceOperator, so the merging operation is eliminated, AnalyticNode with partition_exprs now can parallelized, each AnalyticSinkOperator processes an sorted partition.

### CODE REVIEW

**Comment by satanson** [comment]
run starrocks_be_unittest

File: be/src/exec/pipeline/sort/sort_context.h [hunk 1]
```diff
@@ -232,6 +234,26 @@ class SortContext final : public ContextWithDependency {
 
     size_t _next_output_row = 0;
 };
+class SortContextFactory;
+using SortContextFactoryPtr = std::shared_ptr<SortContextFactory>;
+class SortContextFactory {
+public:
+    SortContextFactory(bool is_merging, int64_t limit, int32_t num_right_sinkers,
```
Comment (kangkaisen) [inline be/src/exec/pipeline/sort/sort_context.h:241]: Why sort context need a Factory? which seems confusing.

File: build.sh [hunk 2]
```diff
@@ -42,10 +42,10 @@ export STARROCKS_HOME=${ROOT}
 . ${STARROCKS_HOME}/env.sh
 
 #build thirdparty libraries if necessary
-if [[ ! -f ${STARROCKS_THIRDPARTY}/installed/lib/mariadb/libmariadbclient.a ]]; then
```
Comment (kangkaisen) [inline build.sh:45]: Revert this change.

File: be/src/exec/vectorized/topn_node.h [hunk 3]
```diff
@@ -39,6 +39,7 @@ class TopNNode final : public ::starrocks::ExecNode {
     std::vector<bool> _is_asc_order;
     std::vector<bool> _is_null_first;
     std::vector<OrderByType> _order_by_types;
+    std::vector<ExprContext*> _analytic_partition_exprs;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/topn_node.h:42]: Comment here.

File: fe/fe-core/src/main/java/com/starrocks/planner/AnalyticPlanner.java [hunk 4]
```diff
@@ -502,6 +502,14 @@ private PlanNode createSortGroupPlan(PlanNode root, SortGroup sortGroup,
                     windowGroup.physicalOutputTuple, windowGroup.logicalToPhysicalSmap,
                     partitionByEq, orderByEq, bufferedTupleDesc);
             node.init(analyzer);
+            // In old planner
+            // Add partition exprs of AnalyticEvalNode to SortNode, it is used in pipeline execution engine
+            // to eliminate time-consuming LocalMergeSortSourceOperator and parallelize AnalyticNode.
+            if (newRoot instanceof SortNode) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/planner/AnalyticPlanner.java:508]: Don't need change old planner.

File: fe/fe-core/src/main/java/com/starrocks/planner/SortNode.java [hunk 5]
```diff
@@ -56,6 +58,7 @@
     private long offset;
     // if true, the output of this node feeds an AnalyticNode
     private boolean isAnalyticSort;
+    private List<Expr> analyticPartitionExprs = Collections.emptyList();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/planner/SortNode.java:61]: Add comment

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 6]
```diff
@@ -1729,6 +1729,15 @@ public PlanFragment visitPhysicalAnalytic(OptExpression optExpr, ExecPlan contex
                 analyticEvalNode.getConjuncts()
                         .add(ScalarOperatorToExpr.buildExecExpression(predicate, formatterContext));
             }
+            // In new planner
+            // Add partition exprs of AnalyticEvalNode to SortNode, it is used in pipeline execution engine
+            // to eliminate time-consuming LocalMergeSortSourceOperator and parallelize AnalyticNode.
+            PlanNode root = inputFragment.getPlanRoot();
+            if (root instanceof SortNode) {
+                SortNode sortNode = (SortNode) root;
+                sortNode.setIsAnalyticSort(true);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1738]: Needn't call  sortNode.setIsAnalyticSort(true);

File: be/src/exec/pipeline/sort/sort_context.h (hunk 1 above)
Comment (satanson) [reply be/src/exec/pipeline/sort/sort_context.h:241]: Sort context is shared by PartitionSortSinkOperator and LocalMergeSortSourceOperator.

File: build.sh (hunk 2 above)
Comment (satanson) [reply build.sh:45]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java (hunk 6 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java]: ok

File: fe/fe-core/src/main/java/com/starrocks/planner/AnalyticPlanner.java (hunk 4 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/AnalyticPlanner.java]: ok

**Review by ZiheLiu — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by Pslydhh — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]
